src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/MessageInterception/InterceptionMiddleLayer2.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+
+namespace Microsoft.VisualStudio.LanguageServer.ContainedLanguage.MessageInterception;
+
+public class InterceptionMiddleLayer2 : ILanguageClientMiddleLayer2
+{
+    private readonly InterceptorManager _interceptorManager;
+    private readonly string _contentType;
+
+    /// <summary>
+    /// Create the middle layer
+    /// </summary>
+    /// <param name="interceptorManager">Interception manager</param>
+    /// <param name="contentType">The content type name of the language for the ILanguageClient using this middle layer</param>
+    public InterceptionMiddleLayer2(InterceptorManager interceptorManager, string contentType)
+    {
+        _interceptorManager = interceptorManager ?? throw new ArgumentNullException(nameof(interceptorManager));
+        _contentType = !string.IsNullOrEmpty(contentType) ? contentType : throw new ArgumentException("Cannot be empty", nameof(contentType));
+    }
+
+    public bool CanHandle(string methodName)
+    {
+        return _interceptorManager.HasInterceptor(methodName, _contentType);
+    }
+
+    public async Task HandleNotificationAsync<TMessage>(string methodName, TMessage message, Func<TMessage, Task> sendNotification)
+    {
+        var payload = message;
+        if (CanHandle(methodName))
+        {
+            payload = await _interceptorManager.ProcessInterceptorsAsync(methodName, payload, _contentType, CancellationToken.None);
+        }
+
+        if (payload is not null)
+        {
+            // this completes the handshake to give the payload back to the client.
+            await sendNotification(payload);
```

Comment:
Thanks. Fortunately I don't think there are any external callers to this so easy enough to change.

Reply:
> you need to invoke sendNotification synchronously

I'm just properly parsing this now, and I'm confused. If it needs to be sync, why didn't it change to `Action<TMessage>` instead of `Func<TMessage, Task>`? Whats the fallout of leaving this as is? Web Tools has an interception handlers for the publish diagnostics notification that does async work

Reply:
This will throw an exception:

await Task.Delay(payload)
await sendNotification(payload);

Instead, everything before sendNotification needs to run synchronously. e.g.

var updatedPayload = ProcessInterceptors(payload);
await sendNotification(payload);

Otherwise if the middle layer is slow to process notifications they could be sent to the server out-of-order and cause non-deterministic behaviors.

Reply:
Thanks, that makes more sense. Does the strictness have to change because of the JToken change, or is it something else?

I have not much context here, but I think this will mean Web Tools can't move to the new API easily. They have an async call in their interceptor, [here](https://devdiv.visualstudio.com/DevDiv/_git/WebTools?path=/src/Languages/LanguageServer/Delegation/MessageInterception/HtmlPublishDiagnosticsInterceptor.cs&version=GBmain&line=52&lineEnd=53&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) that I assume calls out to their OOP server process. I have no idea how easy or hard that is to make sync, and I'm not sure anyone else does either at this point!

Reply:
Is that the only async caller? I could bypass the check for PublishDiagnostics only until they migrate if so...

Also, if they could move to pull diagnostics we wouldn't need the publish diagnostics middleware.

Reply:
or we could potentially bypass the check for notifications going from server->client.

Reply:
Definitely moving to pull diagnostics would be the ideal, but I've no idea if that is realistic. For all I know maybe they already have and the web tools code is never called, but I'm trying to approach this purely from the point of view of owning a public API, and wanting that API to makes sense and not break anything. Personally, it seems to me like bypassing the check for notifications makes sense, since I'm not clear on how ordering should be expected to work with server to client notifications. I guess with that, we would just need separate APIs for notifications and requests, one async, one sync.